File: ios/WhoAmITodayWidgetExtension/NetworkManager.swift
import Foundation

class NetworkManager {
    static let shared = NetworkManager()
    // TODO(Gina): Update with actual API base URL
    private let baseURL = "https://whoami-admin-group.gina-park.site"

    // DEBUG: Store last API status for display
    static var lastProfileStatus: Int = 0
    static var lastFriendsStatus: Int = 0
    static var lastPlaylistStatus: Int = 0
    static var lastQuestionsStatus: Int = 0
    static var lastError: String = ""
    static var questionsDebugInfo: String = ""  // NEW: detailed question fetch info

    func fetchWidgetData() async throws -> WidgetData {
        print("[Widget] fetchWidgetData called")
        print("[Widget] isAuthenticated: \(SharedDataManager.shared.isAuthenticated)")
        print("[Widget] csrfToken exists: \(SharedDataManager.shared.csrfToken != nil)")
        print("[Widget] accessToken exists: \(SharedDataManager.shared.accessToken != nil)")

        guard SharedDataManager.shared.isAuthenticated,
              let csrfToken = SharedDataManager.shared.csrfToken,
              let accessToken = SharedDataManager.shared.accessToken else {
            print("[Widget] Not authenticated - throwing error")
            throw NetworkError.notAuthenticated
        }

        print("[Widget] Starting API calls...")

        // Fetch all data in parallel
        async let myProfile = fetchMyProfile(csrf: csrfToken, token: accessToken)
        async let friends = fetchFriendsWithUpdates(csrf: csrfToken, token: accessToken)
        async let playlists = fetchPlaylistFeed(csrf: csrfToken, token: accessToken)
        async let dailyQuestions = fetchDailyQuestions(csrf: csrfToken, token: accessToken)

        let profileResult = try? await myProfile
        let friendsResult = (try? await friends) ?? []
        let playlistsResult = (try? await playlists) ?? []
        var questionsResult = try? await dailyQuestions

        print("[Widget] Profile result: \(profileResult != nil ? "success" : "nil")")
        print("[Widget] Friends count: \(friendsResult.count)")
        print("[Widget] Playlists count: \(playlistsResult.count)")
        print("[Widget] Daily questions result: \(questionsResult != nil ? "success (\(questionsResult?.count ?? 0) questions)" : "nil")")

        // Fallback to general questions if daily questions failed or empty
        if questionsResult == nil || questionsResult?.isEmpty == true {
            let dailyWasEmpty = questionsResult?.isEmpty == true
            NetworkManager.questionsDebugInfo = "Daily: \(dailyWasEmpty ? "empty[]" : "failed")"
            print("[Widget] Daily questions empty or failed, trying general questions...")

            do {
                questionsResult = try await fetchGeneralQuestions(csrf: csrfToken, token: accessToken)
                print("[Widget] General questions fallback result: success (\(questionsResult?.count ?? 0) questions)")
                NetworkManager.questionsDebugInfo += " → Gen: OK (\(questionsResult?.count ?? 0))"
            } catch {
                print("[Widget] General questions error: \(error)")
                NetworkManager.questionsDebugInfo += " → Gen: ERR[\(error)]"
                questionsResult = nil
            }
        } else {
            NetworkManager.questionsDebugInfo = "Daily: OK (\(questionsResult?.count ?? 0))"
        }

        if let firstQuestion = questionsResult?.first {
            print("[Widget] First question content: \(firstQuestion.content)")
            NetworkManager.questionsDebugInfo += " | Content: \(firstQuestion.content.prefix(50))"
        } else {
            NetworkManager.questionsDebugInfo += " | NO QUESTION FOUND"
        }

        // Use myCheckIn from SharedDataManager (synced from RN with album image) if available,
        // otherwise fall back to API data
        var myCheckInData = SharedDataManager.shared.myCheckIn ?? profileResult?.checkIn
        let source = SharedDataManager.shared.myCheckIn != nil ? "SharedDataManager" : "API"
        print("[Widget] fetchWidgetData: myCheckIn from \(source), mood: \(myCheckInData?.mood ?? "nil")")

        // Fetch album image from Spotify if we have a trackId but no albumImageUrl
        if var checkIn = myCheckInData,
           !checkIn.trackId.isEmpty,
           (checkIn.albumImageUrl == nil || checkIn.albumImageUrl?.isEmpty == true),
           SpotifyManager.shared.isConfigured {
            print("[Widget] Fetching album image from Spotify for trackId: \(checkIn.trackId)")
            if let track = try? await SpotifyManager.shared.getTrack(trackId: checkIn.trackId) {
                checkIn = MyCheckIn(
                    id: checkIn.id,
                    isActive: checkIn.isActive,
                    createdAt: checkIn.createdAt,
                    mood: checkIn.mood,
                    socialBattery: checkIn.socialBattery,
                    description: checkIn.description,
                    trackId: checkIn.trackId,
                    albumImageUrl: track.albumImageUrl
                )
                myCheckInData = checkIn
                print("[Widget] Album image fetched: \(track.albumImageUrl ?? "nil")")
            }
        }

        let data = WidgetData(
            myCheckIn: myCheckInData,
            friendsWithUpdates: friendsResult,
            sharedPlaylists: playlistsResult,
            questionOfDay: questionsResult?.first,
            lastUpdated: Date()
        )

        // Cache the data
        SharedDataManager.shared.cachedWidgetData = data

        print("[Widget] Data cached successfully")

        return data
    }

    /// Validate that stored tokens are still valid (API returns 200 for /api/user/me/profile).
    /// Use this so widget shows "Please Sign in" when user logged out in WebView but tokens remain in App Group.
    func validateToken() async -> Bool {
        guard let csrf = SharedDataManager.shared.csrfToken,
              let token = SharedDataManager.shared.accessToken,
              !csrf.isEmpty, !token.isEmpty else {
            return false
        }
        do {
            _ = try await fetchMyProfile(csrf: csrf, token: token)
            NSLog("[Widget] validateToken: 200 OK, token valid")
            return true
        } catch {
            NSLog("[Widget] validateToken failed (treat as logged out): %@", String(describing: error))
            return false
        }
    }

    // MARK: - Fetch My Profile (/api/user/me/profile)
    private func fetchMyProfile(csrf: String, token: String) async throws -> MyProfileResponse {
        let url = URL(string: "\(baseURL)/api/user/me/profile")!
        var request = URLRequest(url: url)
        request.setValue(csrf, forHTTPHeaderField: "X-CSRFToken")
        // Cookie header should contain both csrftoken and access_token
        request.setValue("csrftoken=\(csrf); access_token=\(token)", forHTTPHeaderField: "Cookie")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            NetworkManager.lastError = "Net: \(error.localizedDescription.prefix(30))"
            throw error
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            print("[Widget] fetchMyProfile: Invalid response type")
            throw NetworkError.invalidResponse
        }

        NetworkManager.lastProfileStatus = httpResponse.statusCode
        print("[Widget] fetchMyProfile status: \(httpResponse.statusCode)")

        guard httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            NetworkManager.lastError = "Profile: \(httpResponse.statusCode)"
            print("[Widget] fetchMyProfile failed: \(body)")
            throw NetworkError.invalidResponse
        }

        do {
            return try JSONDecoder().decode(MyProfileResponse.self, from: data)
        } catch {
            NetworkManager.lastError = "Profile decode: \(error.localizedDescription)"
            throw error
        }
    }

    // MARK: - Fetch Friends with Updates (/api/user/friends/updates/)
    private func fetchFriendsWithUpdates(csrf: String, token: String) async throws -> [FriendUpdate] {
        let url = URL(string: "\(baseURL)/api/user/friends/updates/")!
        var request = URLRequest(url: url)
        request.setValue(csrf, forHTTPHeaderField: "X-CSRFToken")
        request.setValue("csrftoken=\(csrf); access_token=\(token)", forHTTPHeaderField: "Cookie")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            NetworkManager.lastError = "Net Fr: \(error.localizedDescription.prefix(25))"
            throw error
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            print("[Widget] fetchFriendsWithUpdates: Invalid response type")
            throw NetworkError.invalidResponse
        }

        NetworkManager.lastFriendsStatus = httpResponse.statusCode
        print("[Widget] fetchFriendsWithUpdates status: \(httpResponse.statusCode)")

        guard httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            NetworkManager.lastError = "Friends: \(httpResponse.statusCode)"
            print("[Widget] fetchFriendsWithUpdates failed: \(body)")
            throw NetworkError.invalidResponse
        }

        do {
            // Try paginated response first, fallback to direct array
            if let paginatedResponse = try? JSONDecoder().decode(PaginatedResponse<FriendUpdate>.self, from: data) {
                return paginatedResponse.results
            }
            // Fallback: API returns array directly
            return try JSONDecoder().decode([FriendUpdate].self, from: data)
        } catch {
            NetworkManager.lastError = "Friends decode: \(error.localizedDescription)"
            throw error
        }
    }

    // MARK: - Fetch Playlist Feed (/api/playlist/feed)
    private func fetchPlaylistFeed(csrf: String, token: String) async throws -> [PlaylistSong] {
        let url = URL(string: "\(baseURL)/api/playlist/feed")!
        var request = URLRequest(url: url)
        request.setValue(csrf, forHTTPHeaderField: "X-CSRFToken")
        request.setValue("csrftoken=\(csrf); access_token=\(token)", forHTTPHeaderField: "Cookie")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            NetworkManager.lastError = "Net Pl: \(error.localizedDescription.prefix(25))"
            throw error
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            print("[Widget] fetchPlaylistFeed: Invalid response type")
            throw NetworkError.invalidResponse
        }

        NetworkManager.lastPlaylistStatus = httpResponse.statusCode
        print("[Widget] fetchPlaylistFeed status: \(httpResponse.statusCode)")

        guard httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            NetworkManager.lastError = "Playlist: \(httpResponse.statusCode)"
            print("[Widget] fetchPlaylistFeed failed: \(body)")
            throw NetworkError.invalidResponse
        }

        do {
            // API returns paginated response with results array
            let paginatedResponse = try JSONDecoder().decode(PaginatedResponse<PlaylistSong>.self, from: data)
            return paginatedResponse.results
        } catch {
            NetworkManager.lastError = "Playlist decode: \(error.localizedDescription)"
            throw error
        }
    }

    // MARK: - Fetch Daily Questions (/api/qna/questions/daily/)
    private func fetchDailyQuestions(csrf: String, token: String) async throws -> [QuestionOfDay] {
        let url = URL(string: "\(baseURL)/api/qna/questions/daily/")!
        var request = URLRequest(url: url)
        request.setValue(csrf, forHTTPHeaderField: "X-CSRFToken")
        request.setValue("csrftoken=\(csrf); access_token=\(token)", forHTTPHeaderField: "Cookie")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            print("[Widget] fetchDailyQuestions network error: \(error)")
            throw error
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            print("[Widget] fetchDailyQuestions: Invalid response type")
            throw NetworkError.invalidResponse
        }

        NetworkManager.lastQuestionsStatus = httpResponse.statusCode
        print("[Widget] fetchDailyQuestions status: \(httpResponse.statusCode)")

        guard httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            NetworkManager.lastError = "Questions: \(httpResponse.statusCode)"
            print("[Widget] fetchDailyQuestions failed: \(body)")
            throw NetworkError.invalidResponse
        }

        // Log raw response for debugging
        if let responseBody = String(data: data, encoding: .utf8) {
            print("[Widget] Daily questions raw response (first 500 chars): \(responseBody.prefix(500))")
        }

        do {
            // API returns an array of questions directly
            let questions = try JSONDecoder().decode([QuestionOfDay].self, from: data)
            print("[Widget] ✅ Successfully decoded \(questions.count) daily questions")
            if questions.isEmpty {
                print("[Widget] ⚠️ Daily questions array is EMPTY!")
            }
            return questions
        } catch {
            print("[Widget] ❌ fetchDailyQuestions decode error: \(error)")
            print("[Widget] Error details: \(error.localizedDescription)")
            throw error
        }
    }

    // MARK: - Fetch General Questions (/api/qna/questions/) - Fallback
    private func fetchGeneralQuestions(csrf: String, token: String) async throws -> [QuestionOfDay] {
        let url = URL(string: "\(baseURL)/api/qna/questions/?page=1")!
        var request = URLRequest(url: url)
        request.setValue(csrf, forHTTPHeaderField: "X-CSRFToken")
        request.setValue("csrftoken=\(csrf); access_token=\(token)", forHTTPHeaderField: "Cookie")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            print("[Widget] fetchGeneralQuestions network error: \(error)")
            throw error
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            print("[Widget] fetchGeneralQuestions: Invalid response type")
            throw NetworkError.invalidResponse
        }

        print("[Widget] fetchGeneralQuestions status: \(httpResponse.statusCode)")

        guard httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            NetworkManager.lastError = "General Q: \(httpResponse.statusCode)"
            print("[Widget] fetchGeneralQuestions failed: \(body)")
            throw NetworkError.invalidResponse
        }

        // Log raw response for debugging
        if let responseBody = String(data: data, encoding: .utf8) {
            print("[Widget] General questions raw response (first 500 chars): \(responseBody.prefix(500))")
        }

        do {
            // API returns paginated response with QuestionGroup array
            let paginatedResponse = try JSONDecoder().decode(PaginatedResponse<QuestionGroup>.self, from: data)
            print("[Widget] ✅ General questions decoded: \(paginatedResponse.results.count) groups")

            if paginatedResponse.results.isEmpty {
                print("[Widget] ⚠️ Results array is EMPTY!")
                return []
            }

            // Get first question from first group
            if let firstGroup = paginatedResponse.results.first {
                print("[Widget] First group date: \(firstGroup.date), questions count: \(firstGroup.questions.count)")

                if firstGroup.questions.isEmpty {
                    print("[Widget] ⚠️ First group has NO questions!")
                    return []
                }

                if let firstQuestion = firstGroup.questions.first {
                    print("[Widget] ✅ Using fallback question - ID: \(firstQuestion.id), Content: \(firstQuestion.content)")
                    return [firstQuestion.toQuestionOfDay()]
                }
            }

            print("[Widget] ⚠️ Could not extract first question from results")
            return []
        } catch {
            print("[Widget] ❌ fetchGeneralQuestions decode error: \(error)")
            print("[Widget] Error details: \(error.localizedDescription)")
            throw error
        }
    }

    enum NetworkError: Error {
        case notAuthenticated
        case invalidResponse
    }
}

File: android/app/src/main/java/com/whoami/today/app/widget/NetworkManager.java
package com.whoami.today.app.widget;

import android.content.Context;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class NetworkManager {
    private static final String TAG = "NetworkManager";
    private static final String BASE_URL = "https://whoami-admin-group.gina-park.site";

    /**
     * Fetch full widget data from API (same as iOS getTimeline -> fetchWidgetData).
     * Saves to SharedPreferences so WidgetUpdateService can use it.
     */
    public static boolean fetchWidgetData(Context context) {
        String csrfToken = SharedPrefsHelper.getCsrfToken(context);
        String accessToken = SharedPrefsHelper.getAccessToken(context);
        if (csrfToken == null || accessToken == null || csrfToken.isEmpty() || accessToken.isEmpty()) {
            Log.d(TAG, "fetchWidgetData: No auth tokens");
            return false;
        }
        try {
            String cookie = "csrftoken=" + csrfToken + "; access_token=" + accessToken;
            JSONObject existing = new JSONObject();
            String existingJson = SharedPrefsHelper.getWidgetData(context);
            if (existingJson != null && !existingJson.isEmpty()) {
                try {
                    existing = new JSONObject(existingJson);
                } catch (Exception ignored) {}
            }
            JSONObject root = new JSONObject();
            if (existing.has("my_check_in") && !existing.isNull("my_check_in")) {
                root.put("my_check_in", existing.get("my_check_in"));
            } else {
                JSONObject profile = getJson(context, BASE_URL + "/api/user/me/profile", cookie);
                if (profile != null && profile.has("check_in") && !profile.isNull("check_in")) {
                    root.put("my_check_in", profile.getJSONObject("check_in"));
                }
            }
            JSONArray friends = fetchFriendsWithUpdates(context, cookie);
            root.put("friends_with_updates", friends != null ? friends : new JSONArray());
            JSONArray playlists = fetchPlaylistFeed(context, cookie);
            root.put("shared_playlists", playlists != null ? playlists : new JSONArray());
            WidgetData.QuestionOfDay question = fetchFirstDailyQuestion(context);
            if (question != null) {
                JSONObject qJson = new JSONObject();
                qJson.put("id", question.id);
                qJson.put("question", question.content);
                qJson.put("content", question.content);
                qJson.put("created_at", question.createdAt != null ? question.createdAt : "");
                root.put("question_of_day", qJson);
            }
            SharedPrefsHelper.saveWidgetData(context, root.toString());
            Log.d(TAG, "fetchWidgetData: saved friends=" + (friends != null ? friends.length() : 0) + " playlists=" + (playlists != null ? playlists.length() : 0));
            return true;
        } catch (Exception e) {
            Log.e(TAG, "fetchWidgetData error", e);
            return false;
        }
    }

    private static JSONObject getJson(Context context, String urlString, String cookie) {
        try {
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("X-CSRFToken", SharedPrefsHelper.getCsrfToken(context));
            conn.setRequestProperty("Cookie", cookie);
            conn.setConnectTimeout(10000);
            conn.setReadTimeout(10000);
            if (conn.getResponseCode() == 200) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) sb.append(line);
                reader.close();
                conn.disconnect();
                return new JSONObject(sb.toString());
            }
            conn.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "getJson " + urlString + ": " + e.getMessage());
        }
        return null;
    }

    private static JSONArray fetchFriendsWithUpdates(Context context, String cookie) {
        try {
            URL url = new URL(BASE_URL + "/api/user/friends/updates/");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("X-CSRFToken", SharedPrefsHelper.getCsrfToken(context));
            conn.setRequestProperty("Cookie", cookie);
            conn.setConnectTimeout(10000);
            conn.setReadTimeout(10000);
            int code = conn.getResponseCode();
            Log.d(TAG, "fetchFriendsWithUpdates: HTTP " + code);
            if (code == 200) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) sb.append(line);
                reader.close();
                String body = sb.toString();
                conn.disconnect();
                JSONArray arr;
                if (body.startsWith("[")) {
                    arr = new JSONArray(body);
                } else {
                    JSONObject paginated = new JSONObject(body);
                    arr = paginated.has("results") ? paginated.getJSONArray("results") : new JSONArray();
                }
                Log.d(TAG, "fetchFriendsWithUpdates: parsed count=" + arr.length());
                return arr;
            }
            conn.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "fetchFriendsWithUpdates: " + e.getMessage());
        }
        return null;
    }

    private static JSONArray fetchPlaylistFeed(Context context, String cookie) {
        try {
            URL url = new URL(BASE_URL + "/api/playlist/feed");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("X-CSRFToken", SharedPrefsHelper.getCsrfToken(context));
            conn.setRequestProperty("Cookie", cookie);
            conn.setConnectTimeout(10000);
            conn.setReadTimeout(10000);
            int code = conn.getResponseCode();
            Log.d(TAG, "fetchPlaylistFeed: HTTP " + code);
            if (code == 200) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) sb.append(line);
                reader.close();
                JSONObject paginated = new JSONObject(sb.toString());
                conn.disconnect();
                JSONArray arr = paginated.has("results") ? paginated.getJSONArray("results") : new JSONArray();
                Log.d(TAG, "fetchPlaylistFeed: parsed count=" + arr.length());
                return arr;
            }
            conn.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "fetchPlaylistFeed: " + e.getMessage());
        }
        return null;
    }

    public static WidgetData.QuestionOfDay fetchFirstDailyQuestion(Context context) {
        // Try daily questions first
        WidgetData.QuestionOfDay question = fetchFromDailyQuestions(context);

        // Fallback to general questions if daily failed or is null
        if (question == null) {
            Log.d(TAG, "Daily questions failed or empty, trying general questions...");
            question = fetchFromGeneralQuestions(context);
        }

        return question;
    }

    private static WidgetData.QuestionOfDay fetchFromDailyQuestions(Context context) {
        try {
            String csrfToken = SharedPrefsHelper.getCsrfToken(context);
            String accessToken = SharedPrefsHelper.getAccessToken(context);

            if (csrfToken == null || accessToken == null) {
                Log.e(TAG, "No auth tokens available");
                return null;
            }

            URL url = new URL(BASE_URL + "/api/qna/questions/daily/");
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("X-CSRFToken", csrfToken);
            connection.setRequestProperty("Cookie", "csrftoken=" + csrfToken + "; access_token=" + accessToken);
            connection.setConnectTimeout(10000);
            connection.setReadTimeout(10000);

            int responseCode = connection.getResponseCode();
            Log.d(TAG, "Daily questions response code: " + responseCode);

            if (responseCode == 200) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                reader.close();

                String responseBody = response.toString();
                Log.d(TAG, "Daily questions raw response (first 500 chars): " + responseBody.substring(0, Math.min(500, responseBody.length())));

                // Parse JSON array and get first question
                JSONArray questionsArray = new JSONArray(responseBody);
                Log.d(TAG, "✅ Daily questions decoded: " + questionsArray.length() + " questions");

                if (questionsArray.length() == 0) {
                    Log.w(TAG, "⚠️ Daily questions array is EMPTY!");
                    return null;
                }

                if (questionsArray.length() > 0) {
                    JSONObject firstQuestion = questionsArray.getJSONObject(0);
                    WidgetData.QuestionOfDay question = WidgetData.QuestionOfDay.fromJson(firstQuestion);
                    Log.d(TAG, "✅ Successfully fetched daily question - ID: " + question.id + ", Content: " + question.content);
                    return question;
                }
            } else {
                Log.e(TAG, "❌ Failed to fetch daily questions: " + responseCode);
            }

            connection.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "Error fetching daily questions: " + e.getMessage());
            e.printStackTrace();
        }

        return null;
    }

    private static WidgetData.QuestionOfDay fetchFromGeneralQuestions(Context context) {
        try {
            String csrfToken = SharedPrefsHelper.getCsrfToken(context);
            String accessToken = SharedPrefsHelper.getAccessToken(context);

            if (csrfToken == null || accessToken == null) {
                Log.e(TAG, "No auth tokens available");
                return null;
            }

            URL url = new URL(BASE_URL + "/api/qna/questions/?page=1");
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("X-CSRFToken", csrfToken);
            connection.setRequestProperty("Cookie", "csrftoken=" + csrfToken + "; access_token=" + accessToken);
            connection.setConnectTimeout(10000);
            connection.setReadTimeout(10000);

            int responseCode = connection.getResponseCode();
            Log.d(TAG, "General questions response code: " + responseCode);

            if (responseCode == 200) {
                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                reader.close();

                String responseBody = response.toString();
                Log.d(TAG, "General questions raw response (first 500 chars): " + responseBody.substring(0, Math.min(500, responseBody.length())));

                // Parse paginated response
                JSONObject paginatedResponse = new JSONObject(responseBody);
                JSONArray resultsArray = paginatedResponse.getJSONArray("results");
                Log.d(TAG, "✅ General questions decoded: " + resultsArray.length() + " groups");

                if (resultsArray.length() == 0) {
                    Log.w(TAG, "⚠️ Results array is EMPTY!");
                    return null;
                }

                if (resultsArray.length() > 0) {
                    // Get first question group
                    JSONObject firstGroup = resultsArray.getJSONObject(0);
                    String groupDate = firstGroup.optString("date", "unknown");
                    JSONArray questionsArray = firstGroup.getJSONArray("questions");
                    Log.d(TAG, "First group date: " + groupDate + ", questions count: " + questionsArray.length());

                    if (questionsArray.length() == 0) {
                        Log.w(TAG, "⚠️ First group has NO questions!");
                        return null;
                    }

                    if (questionsArray.length() > 0) {
                        // Get first question from first group
                        JSONObject firstQuestion = questionsArray.getJSONObject(0);

                        // Convert to QuestionOfDay format
                        WidgetData.QuestionOfDay question = new WidgetData.QuestionOfDay();
                        question.id = String.valueOf(firstQuestion.optInt("id"));
                        question.content = firstQuestion.optString("content", "");
                        question.question = question.content;
                        question.createdAt = firstQuestion.optString("created_at", "");

                        Log.d(TAG, "✅ Using fallback question - ID: " + question.id + ", Content: " + question.content);
                        return question;
                    }
                }

                Log.w(TAG, "⚠️ Could not extract first question from results");
            } else {
                Log.e(TAG, "❌ Failed to fetch general questions: " + responseCode);
            }

            connection.disconnect();
        } catch (Exception e) {
            Log.e(TAG, "Error fetching general questions: " + e.getMessage());
            e.printStackTrace();
        }

        return null;
    }
}
